test_suite/marks.py

```diff
@@ -0,0 +1,159 @@
+from pstats import Stats
+
+import parameterized as pm
+
+from helpers import error, tf_cfg
+from test_suite import tester
+
+
+def retry_if_not_conditions(test):
+    """
+    A decorator to retry the test if the testing conditions are not completed.
+
+    This decorator wraps the test function and provides it with up to three
+    attempts for execution if an exception error.TestConditionsAreNotCompleted is raised.
+
+    For example:
+    @retry_if_not_conditions
+    def test_example(self);
+        # part of test
+        if test_not_condition():
+            raise error.TestConditionsAreNotCompleted
+        # asserts etc.
+
+    The dmesg asserts should use range in the expect argument:
+        self.assertFrangWarning(warning=warning, expected=range(1, 15))
+    """
+
+    def wrapper(self: tester.TempestaTest, *args, **kwargs):
+        for attempt in range(1, 4):
+            try:
+                test(self, *args, **kwargs)
+                return
+            except error.TestConditionsAreNotCompleted:
+                tf_cfg.dbg(1, "Test condition are not completed. Stop services and restart test.")
+                # Stop all services on failure
+                for service in self.get_all_services():
+                    service.stop()
+                # Wait for 1 second before retrying
+                time.sleep(1)
+
+        # If the test fails after 3 attempts, raise an exception
+        raise error.TestConditionsAreNotCompleted(self.id(), attempt)
+
+    # we need to change name of function to work correctly with parametrize
+    wrapper.__name__ = test.__name__
+    return wrapper
+
+
+def deprecated(alt_impl_name):
+    """
+    Decorator to declare a class and its descendants as deprecated.
+    Example ('Foo' is a new alternative implementation which should be used
+    instead):
+
+        @util.deprecated("Foo")
+        class A(...):
+            ...
+    """
+
+    def decorator(cls):
+        def deprecated_new(new_func):
+            def wrap(cls_arg, *args, **kwargs):
+                tf_cfg.dbg(
+                    6,
+                    "%s must be used instead of deprecated %s" % (alt_impl_name, cls_arg.__name__),
+                )
+                return new_func(cls_arg)
+
+            return wrap
+
+        setattr(cls, "__new__", staticmethod(deprecated_new(cls.__new__)))
+        return cls
+
+    return decorator
+
+
+def profiled(func):
+    """
+    Profiling decorator, you can use it as:
+
+        class SomeTester:
+            @util.profiled
+            def test_func(self):
+                ....
+
+    Results will be printed to console.
+    """
+
+    @functools.wraps(func)
+    def wrap(*args, **kwargs):
+        prof = Profile()
+        res = prof.runcall(func, *args, **kwargs)
+        stats = Stats(prof)
+        stats.strip_dirs()
+        stats.sort_stats("tottime")
+        stats.print_stats(20)
+        return res
+
+    return wrap
+
+
+def get_func_name(func, param_num, params):
```

Comment:
My thoughts about the module and the components.

We declared decorators here to provide them for users. For example, `get_func_name` and `get_class_name` are internal functions that some decorators use and I suppose we do not want to use them in somewhere except this module. So, I would suggest to rename them with `_` prefix as a protected method to show a user that is internal component. 

And I would like to suggest to name classes in camel case as in the convention, if you do not have other reasons to not do so.

Reply:
I agree with protected method, I will fix it. About `parametrize` class - I also just repeated the parametrized library. I think, in this case, it will just be more practical - `marks.parametrize` or `marks.Parametrize`

Reply:
IMHO
We do not need to repeat a code style from any library where we use this library - we are creating our own product with our own code style. If we would copy every code style for every part where we use a different library it would be chaos. 
If we chose to use class names in camel style let's do it everywhere, it isa about the consistency.

Reply:
Yes, I agree. I changed function names and class names.